File: nvpl_blas/c/srot.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_srot
 *
 ******************************************************************************/
#include "example_helper.h"
#include <math.h> // sqrt

int main() {
    nvpl_int_t N = 5;
    float * X;
    float * Y;
    float s = 0.5f;
    float c = sqrt(1 - s * s);
    nvpl_int_t incX = 1;
    nvpl_int_t incY = 1;

    printf("\nExample: cblas_srot for applying Givens rotation matrix\n");
    printf("#### args: n=%" PRId64 ", incx=%" PRId64 ", incy=%" PRId64 ", c=%f, s=%f\n", (int64_t)N,
            (int64_t)incX, (int64_t)incY, c, s);

    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);
    nvpl_int_t len_y = 1 + (N - 1) * labs(incY);
    // allocate memory
    X = (float *)malloc(len_x * sizeof(float));
    Y = (float *)malloc(len_y * sizeof(float));

    // fill data
    fill_svector(X, N, incX);
    fill_svector(Y, N, incY);

    // print input data
    print_svector(X, N, incX, "X");
    print_svector(Y, N, incY, "Y");

    // call cblas_srot
    cblas_srot(N, X, incX, Y, incY, c, s);

    // print result
    printf("\nThe result of cblas_srot:\n");
    print_svector(X, N, incX, "X");
    print_svector(Y, N, incY, "Y");

    // release memory
    free(X);
    free(Y);
    return 0;
}

File: nvpl_blas/f77/srot.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     SROT
!
!******************************************************************************
      program   SROT_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic       sqrt, abs

      integer         n, incx, incy, i
      parameter       (n=5, incx=1, incy=1)
      real            s, c
      parameter       (s=0.5, c=sqrt(1-s*s))
      integer         xsize, ysize
      parameter       (xsize=1+(n-1)*abs(incx),
     &                 ysize=1+(n-1)*abs(incy))
      real            x(xsize), y(ysize)
      data            (x(i), i=1,xsize)/1.0,-3.7,4.3,0.0,5.0/,
     &                (y(i), i=1,ysize)/ysize*2.0/

*       External Subroutines
      external        print_svector
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, incx, incy, c, s
      call print_svector(x,n,incx,'X')
      call print_svector(y,n,incy,'Y')
*
*      Call SROT subroutine
      call SROT(n,x,incx,y,incy,c,s)
*
      print*
      print 101
      call print_svector(x,n,incx,'X')
      call print_svector(y,n,incy,'Y')

      stop
 99   format('Example: SROT for applying Givens rotation matrix')
 100  format('#### args: n=',i1,', incx=',i1,', incy=',i1,
     &       ', c=',f5.3,', s=',f5.3)
 101  format('The result of SROT: ')
      end
